File: api/gads-results.py
from api._lib.blob_io import load_df
from api._lib.core.gads_opportunity import compute_opportunity
from api._lib.handler_utils import BaseHandler, records


class handler(BaseHandler):
    def do_POST(self):
        try:
            body = self._json_body()
            combined_df = load_df(body["combinedDfBlobUrl"])
            gads_df = load_df(body["gadsDfBlobUrl"])
            final_view = compute_opportunity(combined_df, gads_df)

            self._send_json(200, {
                "hasResults": not final_view.empty,
                "preview": records(final_view),
                "totalRows": len(final_view),
                "keywordsWithVolume": (
                    int(final_view["monthly_searches"].notna().sum())
                    if "monthly_searches" in final_view.columns
                    else 0
                ),
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/gads-results.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import { sendError, records } from '../api/_lib/handlerUtils.js'
import { loadRows } from '../api/_lib/sessionStore.js'
import { computeOpportunity } from '../api/_lib/core/gadsOpportunity.js'

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' })

  try {
    const { combinedDfBlobUrl, gadsDfBlobUrl } = req.body as { combinedDfBlobUrl: string; gadsDfBlobUrl: string }

    const combined = await loadRows(combinedDfBlobUrl)
    const gads = await loadRows(gadsDfBlobUrl)
    const finalView = computeOpportunity(combined, gads)

    const withVolume = new Set(finalView.filter(r => r.monthly_searches).map(r => r.keyword))

    res.status(200).json({
      hasResults: finalView.length > 0,
      preview: records(finalView),
      totalRows: finalView.length,
      keywordsWithVolume: withVolume.size,
    })
  } catch (err) {
    sendError(res, err)
  }
}
